File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:

        m = len(grid)
        n = len(grid[0])

        max_count = 0
        count = [0]

        dir = [(1,0), (0,1), (0,-1), (-1,0)]

        def dfs(r,c):
            if r<0 or r>m-1 or c<0 or c>n-1 or grid[r][c] != 1:
                return

            grid[r][c] = 0
            count[0] += 1

            for dR, dC in dir:
                dfs(r+dR, c+dC)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    dfs(i,j)
                    max_count = max(max_count, count[0])
                    count[0] = 0

        return max_count
        

File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {

        int m = grid.length;
        int n = grid[0].length;
        int max_area = 0;
        

        for (int i=0; i<m; i++)
         for (int j = 0; j<n; j++)
        {
            if (grid[i][j] == 1)
            {
                int area = dfs(i, j, m, n, grid);
                max_area = Math.max(max_area, area);
            }
        }
        
        return max_area;
    }

     public int dfs(int i, int j, int m, int n, int [][] grid)
        {
            //boundary conditions
            if (i < 0 || i > m-1 || j < 0 || j > n-1 || grid[i][j] == 0)
            return 0; //wothless

            else {

                //first flood the cell
                grid[i][j] = 0;

                return 1+ dfs(i+1,j, m, n, grid)+ dfs(i, j-1, m, n, grid)+ dfs(i, j+1, m, n, grid)+dfs(i-1, j, m, n, grid);//do dfs on its 4 neighbors
            }

        }
}